### DISCUSSION TITLE
Anyway to concatenate env variables for schema.prisma file?

### DISCUSSION THREAD

**Discussion by heymartinadams** [prisma/prisma#4795] [Q&A] [answered]
We’re using several staging environments (`stg_001`, `stg_002`, etc) and the only way we can have those in place (with Vercel, which only let’s us use a single staging environment) is if we have distinct environment variables, e.g.

`DATABASE_URL_STG_001`, `DATABASE_URL_STG_002`, etc.

Any chance I can concatenate the env variable in the `schema.prisma` file?

e.g.

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_" + process.env.CUSTOM_ENV)
}
```

Wonder if this could be related to https://github.com/prisma/prisma/discussions/4783

ps: I use a secrets manager so `dotenv-expand` won’t be useful (as mentioned [here](https://www.prisma.io/docs/concepts/components/prisma-schema#accessing-environment-variables-from-the-schema)).

**Comment by pantharshit00** [comment, marked as answer]
@heymartinadams This is not possible right now. Feature request which is tracking this: https://github.com/prisma/prisma/issues/2559

**Comment by janpio** [comment]
Why does the usage of a screts manager make the usage of the integrated dotenv-expand functionality impossible?

**Reply by heymartinadams** [reply]
Hi @janpio — don’t recall exactly why `dotenv-expand` didn’t work for me at the time. But I was able to make it work with a single variable `DATABASE_URL` for all environments. Our secrets manager simply fills in different values into that variable, depending on which environment. Personally, don’t need the above-mentioned feature.

**Reply by nickfreemandesign** [reply]
@janpio We are also using AWS Secrets Manager along with ECS, here is why [#2559](https://github.com/prisma/prisma/issues/2559) is a big deal to us:

Secrets Manager holds our DB credentials as individual key:value pairs (`DB_PASSWORD`, `DB_HOST`, `DB_PORT`, etc) and [configured for rotation](https://docs.aws.amazon.com/secretsmanager/latest/userguide/rotating-secrets.html) as a best practice. Meanwhile, our Prisma Client runs inside an ECS cluster which allows for [injecting secrets](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data-secrets.html) but only one ARN at a time. This results in a bottleneck at AWS Secrets and ECS where concatenating variables is not possible unless a _new_ key:value pair is manually created in Secrets Manager, where the concatenation is done. That strategy however, prevents secret rotation which is more of a priority than sticking with Prisma. 

There _may_ be a way to concatenate multiple ARNs, but AFAIK it's not possible especially when considering rotating secrets. The road of least resistance then, leads to Prisma allowing for the following syntax to be possible.

```prisma
datasource db {
  provider = "postgresql"
  url      = env(`postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}`)
}
```

**Reply by janpio** [reply]
Thanks. Feel free to copy the full explanation to #2559 as well.

Honestly I still do not fully understand yet why https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables would not work in such a scenario though. For me this seems functionally equivalent to the snippet you posted, if though with a detour via a `.env` file hat concatenates into one env var.

(Does not change that us implementing #2559 of course might be the better solution.)

**Reply by nickfreemandesign** [reply]
Sure thing @janpio, TL;DR `.env` is of no use when working with Secrets Manager, and therefore neither is dotenv-expand.

Not sure if this helps clarify things but think of AWS Secrets Manager as a _replacement_ to `.env` files altogether. It's essentially a managed `.env` service that allows for secret rotation, encryption via KMS, and of course all of the trappings afforded through IAM policy provisioning. Because [dotenv-expand](https://github.com/motdotla/dotenv-expand) is a tool which works with `.env` files, it has no use when it comes to AWS Secrets Manager. I would love for AWS Secrets Manager to build in a dotenv-expand functionality where secrets can reference each other, but for now AFAIK that is not possible. Because of this (and the strength in security provided by Secrets Manager), the path of least resistance here is to allow the syntax within the Prisma Client.

I'd be glad to dig in further if there is more confusion to clear up. Either way, thanks for seeking to understand

**Reply by janpio** [reply]
Let's go a bit further: I understood AWS Secrets Manager would be executed before your app runs, and load secrets from an external source and put it into the environment. Then when Prisma CLI or Client is launched, it reads a `.env` file (if present) and does things via `dotenv-expand` in there. One of those things can be to combine/expand existing env vars (which could have been created via AWS Secrets Manager) together to create a new env var and put that into the environment (which Prisma CLI then can read). Or not?

**Reply by nickfreemandesign** [reply]
> One of those things can be to combine/expand existing env vars (which could have been created via AWS Secrets Manager)

What I'm understanding you to say is: In between Secrets Manager and loading your app containing Prisma Client, you can create a process which extracts from Secrets Manager into a `.env` file, and then use the existing functionality to get what we want. 

If I got that correctly, then yes that is possible but a bit of a merry-go-round implementation to get there. The challenge would also then become triggering this process within our CI/CD pipeline, and granting access to the AWS SDK and IAM permissions to read from Secrets Manager then write back to the CodeBuild container where the code is built.  All in all, not a deal breaker, but not ideal. 

We may work in this direction as time progresses, but I think PSL would stand to benefit from the suggested syntax– if not for this reason, likely for others. Either way, I'm feeling like you understand the challenge and benefit of the proposed syntax for those who use a managed service to store their secrets.

**Reply by janpio** [reply]
> I'm feeling like you understand the challenge and benefit of the proposed syntax for those who use a managed service to store their secrets.

I absolutely do, and wish I could give you a clear ETA when we will have time to deal with this.

Let me rephrase my understanding, to make it clearer:
- In my understanding (and unfortuantely not experience, so I am going from your description and a skim of the docs) AWS Secrets Manager pulls in key-value pairs (e.g. for DB_USER, DB_PASSWORD, DB_HOST and DB_PORT) from a secure location and puts them into your environment where you run your application (e.g. Lamba).
- Your application includes Prisma Client, which reads `.env` when instantiated.
- In `.env` you should be able to "build" new key-value pairs with the pieces that are already in your environment (via AWS Secrets Manager): `DATABASE_URL=postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}`
- After it did that, it puts the result into the environment as well - and then when the Prisma Schema is parsed it fills `DATABASE_URL` with that value.

**Reply by nickfreemandesign** [reply]
Rephrasing the last 3 bullets:

- Your application includes Prisma Client, which **reads directly from Secrets Manager** when instantiated.
- **In `schema.prisma`** you should be able to "build" new key-value pairs with the pieces that are already in your environment (via AWS Secrets Manager): env(postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT})
- ~After it did that, it puts the result into the environment as well - and then~ When the Prisma Schema is parsed it fills **`postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}` instead of `DATABASE_URL`** with that value.

_schema.prisma file:_
```prisma
datasource db {
  provider = "postgresql"
  url      = env(`postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}`)
}
```

**Reply by janpio** [reply]
> ... which reads directly from Secrets Manager when instantiated.

I don't think one can read from Secrets Manager. Secrets Manager only puts the values into the actual environment so that _any_ app can read it from there - one does not need special support for it.

That then leaves that we need to concatenate individual env variables in `schema.prisma` (or `env()` to be more exact) somehow, and that is exactly what we track in https://github.com/prisma/prisma/issues/2559 (but is a bit harder than it looks on first sight because of our architecture).

If you have the capacity, I would appreciate if you could test if the `.env` file approach indeed works with a deployment that uses AWS Secrets Manager.

**Reply by nickfreemandesign** [reply]
> somehow, and that is exactly what we track in https://github.com/prisma/prisma/issues/2559 (but is a bit harder than it looks on first sight because of our architecture)

No doubt! Y'all invented a new language, I can only imagine this is not a trivial task. Kudos to the team

> If you have the capacity, I would appreciate if you could test if the .env file approach indeed works with a deployment that uses AWS Secrets Manager.

I can certainly give that a try. I'll work with my team to prioritize this in our next sprint (early Sept 2022). 

> I don't think one can read from Secrets Manager. Secrets Manager only puts the values into the actual environment so that any app can read it from there - one does not need special support for it.

I'm not totally sure what you mean on this one, but perhaps its clarifying if I show you what I mean by "read from Secrets Manager"

In ECS, containers are defined within "Task Definitions" (~same as a Pod in K8s), where secrets are injected at runtime (just like an `.env` file would behave) via their secret ARN. The image below shows a manually concatenated `DATABASE_URL` so that Prisma will be happy, but the ideal solution involves secret rotation and should be `DB_PASSWORD`, `DB_HOST`, `DB_PORT` etc, where the concatenation happens in `schema.prisma`

![container definition showing secret injection within ECS Task Definition](https://user-images.githubusercontent.com/12129271/184379945-a79ea2ca-54cd-4aac-a35e-c610773eac25.png)

In order to pull off this workaround where I suggest "reading from Secrets Manager", this would be done either using the AWS SDK in a preferred language ([python example reading secrets](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/secrets-manager.html) or by using the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/secretsmanager/get-secret-value.html). In any event, I would describe this as a read operation against Secrets Manager to ultimately generate a `.env` file, and _then_ inject to the container definition from there.

![container definition showing environment variable injection within ECS Task Definition](https://user-images.githubusercontent.com/12129271/184381389-d31fdcee-1dbf-4025-bd78-8940f009d737.png)

Alrighty. Hope this has been useful. Thanks for dialoguing through this with me. I'll report back once our team frees up to implement this workaround 🤙

**Reply by janpio** [reply]
I think that read operation would not be necessary right now:
You are accessing that `DATABASE_URL` (which you created as a workaround) in your Prisma Schema file via `url = env("DATABASE_URL")`, correct? That actually means that AWS Secrets Manager took that value and put it "into the environment". 
If you had the individual pieces (`DB_USER` etc) in there instead you should be able to have a `.env` file with content `DATABASE_URL=postgresql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}` and magically the Prisma Schema would get the concatenated `DATABASE_URL`.

**Reply by janpio** [reply]
Yeah, I have just been told by a colleague that Secrets Manager indeed does _not_ put things into the environment, but you would need to do that manually. Ignore everything I wrote above then - it will not work with AWS Secrets Manager. Sorry for wasting both our time.

**Reply by nickfreemandesign** [reply]
LOL not a waste, all in the name of learning more than yesterday 🦾 . I'll try the workaround script and report back. Happy end-of-week to you. Cheers

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
